test/unit/io/sequence_file/sequence_file_format_embl_test.cpp

```diff
@@ -0,0 +1,573 @@
+            EXPECT_EQ(id, expected_ids[i]);
+            EXPECT_EQ(seq, expected_seqs[i]);
+            EXPECT_TRUE((ranges::equal(seq, expected_seqs[i])));
+            EXPECT_TRUE((ranges::equal(id, expected_ids[i])));
+        }
+    }
+};
+
+TEST_F(read, standard)
+{
+    std::string input
+    {
+        "ID ID1;\tstuff\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT TTTTTTTT        18\n"
+        "//\n"
+        "ID ID2;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT TTTTTTTTTT TTTTTTTTTT TTTTTTTTTT TTTTTTTTTT TTTTTTTTTT 60\n"
+        "TTTTTTTTTT TTTTTTTTTT TT        82\n"
+        "//\n"
+        "ID ID3 lala;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTA        7\n"
+        "//\n"
+    };
+
+    do_read_test(input);
+}
+
+TEST_F(read, whitespace_in_seq)
+{
+    std::string input
+    {
+        "ID ID1;\tstuff\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTT        18\n"
+        "//\n"
+        "ID ID2;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTTTT\nTTTTTTTTTT\vTTTTTTTTTT\tTTTTTTTTTT TTTTTTTTTT 60\n"
+        "TTTTTTTTTT TTTTTTTTTT TT        82\n"
+        "//\n"
+        "ID ID3 lala;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTA        7\n"
+        "//\n"
+    };
+    do_read_test(input);
+}
+
+TEST_F(read, no_id)
+{
+    std::string input
+    {
+        "IK ID1;\tstuff\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTT        18\n"
+        "//\n"
+        "ID ID2;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTTTT\nTTTTTTTTTT\vTTTTTTTTTT\tTTTTTTTTTT TTTTTTTTTT 60\n"
+        "TTTTTTTTTT TTTTTTTTTT TT        82\n"
+        "//\n"
+        "ID ID3 lala;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTA        7\n"
+        "//\n"
+    };
+
+    std::stringstream istream{input};
+    EXPECT_THROW(( format.read(istream, options, seq, id, std::ignore)), parse_error );
+}
+
+TEST_F(read, other_lines)
+{
+    std::string input
+    {
+        "ID ID1;\tstuff\n"
+        "XX\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTT        18\n"
+        "//\n"
+        "ID ID2;\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTTTTT\nTTTTTTTTTT\nTTTTTTTTTT\vTTTTTTTTTT\tTTTTTTTTTT TTTTTTTTTT 60\n"
+        "TTTTTTTTTT TTTTTTTTTT TT        82\n"
+        "//\n"
+        "ID ID3 lala;\n"
+        "AC 2389ß\n"
+        "SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;\n"
+        "  ACGTTTA        7\n"
+        "//\n"
```

Comment:
These would probably be more readable if you use raw string literals:
(6) on https://en.cppreference.com/w/cpp/language/string_literal

Reply:
I think this is more readable, that is how the file would look like for a human, right?

Reply:
```cpp
    std::string input
    {
R"(ID ID1;  stuff
XX
SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;
  ACGTTTTTTT
TTTTTTTT        18
//
ID ID2;
SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;
  ACGTTTTTTT
TTTTTTTTTT
TTTTTTTTTTTTTTTTTTTT    TTTTTTTTTT TTTTTTTTTT 60
TTTTTTTTTT TTTTTTTTTT TT        82
//
ID ID3 lala;
AC 2389ß
SQ Sequence 1859 BP; 609 A; 314 C; 355 G; 581 T; 0 other;
  ACGTTTA        7
//)"
    };
```

would work.
Though I don't know how to write `\v`.

Reply:
But the problem is that I than can't have any tabs before ID ID2; or ID ID3, because I am assuming I am always at the beginning of an ID, if I am not after reading one I get an error. Do you want me to consume this empty fields in the beginning? If not, then using raw string literals would mean to have it look like Enrico posted it, so the strings are not indented... Which would be inconsistent with other files...

Reply:
It's ok that it's not indented, I think it will still be a lot clearer to see that this is "raw text"

Reply:
We tried this in one of @smehringer's snippets.
One problem is that you need to process the raw string to replace the spaces with tabs again.
We did it for a SAM file, where it is relatively easy since you don't have any spaces in the file we used and can just replace consecutive spaces with a tab.
This is quite handy if you want to print the SAM file and then use it since you only need to modify the raw string in one place.

But here we have mixed spaces and tabs in the format, so this might not be really applicable at all.

Reply:
But you can put literal tabs into the raw string, than it works fine, or not?

Reply:
Mhm I guess you could, but then you need to make your editor behave....

And sometimes (like for SAM), you want to use extra spaces to align the columns for readability.

Reply:
@h-2 Just to clarify, should I change this just for this one example or also for all other embl inputs?

Reply:
@h-2 A question, which is more important for SAM than EMBL but concerns the Raw string, what do I do if my line gets too long due to a long sequence, before I could just break the sequence and put some parts in the next line but with a raw string this means I have a new line character in there.

Reply:
Exceeding the line length is ok for raw strings. Raw string are just a ways of clearly communicating to the reader: "THIS IS NOT CPLUSPLUS, its' just a block of data"